Design/histogram_type.md

```diff
@@ -0,0 +1,108 @@
+  - Many units tests of algorithms do not fully test the range of errors that can happen when, e.g., adding two spectra. Doing so would be very hard in the current de-centralized implementation,
+  - Many algorithms need to modify the errors. In some algorithms the unit tests cover the errors, in others they do not.
+
+
+## Design goals
+
+- Provide simple arithmetic operations for histograms.
+  - Arithmetic operations automatically deal with propagation of errors.
+- Hide differences between histogram and distribution data.
+  - Internal storage type could be either, does not need to be known at the interface level?
+  - Arithmetic operations should do the right thing for any data type automatically.
+- Hide error handling.
+  - For simple operations with histograms algorithm authors should not need to know how to propagate errors.
+- Enable unit testing of key concepts in Mantid.
+  - Arithmetic operations with histograms.
+  - Propagation of errors.
+
+
+
+## Implementation
+
+#### Basics
+
+The details of the implementation have not been worked out yet.
+The basic concept seems simple enough:
+
+- Add one class for a histogram with an interface that provides standard operations for histogram data.
+  - Note that we do not introduce a separate type for histogram and distribution data. This distinction would be only internal, and access to either histogram or distribution data is provided via public interface methods. The goal is to eliminate this distinction from as much of the code as possible.
+- Add a few classes for `X`, `Y`, and `E` (etc.) that would mainly be used internally in the histogram class and for direct access to the underlying data. If these inherit from `std::vetor<double>`, they could also be used to keep the legacy interface for `ISpectrum` alive until the roll-out is complete.
+
+#### Details
+
+- It is probably useful to introduce two types for vectors of standard deviation and variance, instead of the current double use of `E`.
+- Similarly we should have distinct internal types for histogram and distribution data, in case our implementation relies on this distinction. Jon pointed out that it may make most sense to default to distribution data, as OpenGenie did by default.
+- The current copy-on-write (COW) mechanism seems useful and should be preserved.
+  - The `typedef` `MantidVecPtr` for a `cow_ptr` is useless and confusing and should be removed.
+  - We should consider extending the COW mechanism and make sure sharing is preserved more than it is now. For example, what can we do to maintain sharing of `X` for all histograms in a workspace where `X` is modified in an identical way? The COW mechanism as it is now should break down in that case. Does it need to be handled at the workspace level, or can we deal with this here? Having parametric `X` could circumvent that in certain cases (see outlook).
```

Comment:
@OwenArnold What you describes works for sharing initially. However, the issue here is the following: Say we have a workspace where `X` is the same for all spectra, this is shared (also in the current implementation). If you now add an offset to all `X` the sharing is broken, since you modify each individual `X` (and thus trigger the copy-on-write).

Regarding your other thoughts:
- Sharing across workspaces: Not sure I understand what you are referring to, is this different from the current case?
- Sharing is `Y` and `E`: This is also done now. The only current shortcoming is that the only way to copy `Y` and `E` while maintaining sharing is currently via the copy-constructor. But in principle it is done and happens in certain algorithms.